### PR TITLE
Upgrades overhaul part 1: EnabledCondition

### PR DESCRIPTION
**Pull request by pchote**
This is the second step towards #12380 (after #12378, which this depends on).

`EnableCondition` is implemented in parallel to the old code for all of our upgrade consumers, assuming / forcing level 0 &rarr; false, level >= 1 &rarr; true.  I have ported TS over to the new system as an example / testcase (omitting upgrade rules so that this doesn't affect other mods).

I'm looking for suggestions on a better name for `EnableCondition`, because this is ambiguous (could mean that the trait is enabled on this condition, or that the trait grants the condition).

### CODE REVIEW

**Comment by pchote** [comment]
it turns out that `2cd1cf5` was making this fail the linter because the `CheckUpgrades` lint actually test checks granted upgrades across *all* actors, which is kind of bogus.  Fixing that is out of scope of this first PR so I have removed the commit.

**Comment by Phrohdoh** [comment]
re: just the name, `RequiresCondition`?

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradableTrait.cs [hunk 1]
```diff
@@ -56,33 +64,63 @@ public virtual void RulesetLoaded(Ruleset rules, ActorInfo ai)
...
-			// Restrict the levels to the allowed range
-			oldLevel = oldLevel.Clamp(0, Info.UpgradeMaxAcceptedLevel);
-			newLevel = newLevel.Clamp(0, Info.UpgradeMaxAcceptedLevel);
-			if (oldLevel == newLevel)
-				return;
+				// Restrict the levels to the allowed range
+				oldLevel = oldLevel.Clamp(0, Info.UpgradeMaxAcceptedLevel);
+				newLevel = newLevel.Clamp(0, Info.UpgradeMaxAcceptedLevel);
+				if (oldLevel == newLevel)
+					return;
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradableTrait.cs]: This til the end of the block could be:

```csharp
if (oldLevel != newLevel)
    IsTraitDisabled = newLevel < Info.UpgradeMinEnabledLevel || newLevel > Info.UpgradeMaxEnabledLevel;
```

This doesn't matter in the grand scheme of course.

**Comment by pchote** [comment]
`RequiresCondition` sounds good to me.

File: mods/ts/sequences/misc.yaml [hunk 2]
```diff
@@ -62,9 +62,12 @@ crate-effects:
 	levelup: veteran
 
 rank:
-	rank: pips
+	veteran: pips
 		Start: 7
-		Length: 2
+		Length: 1
```
Comment (abcdefg30) [inline mods/ts/sequences/misc.yaml]: Isn't `Length: 1` already the default? https://github.com/OpenRA/OpenRA/blob/bea2b5cc91c3992f47aab955b94888b769855c9e/OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs#L180

**Comment by reaperrr** [comment]
Needs rebase.

**Comment by pchote** [comment]
Rebased, but still needs the tweaks discussed above.

I'm also thinking that this should include upgrade rules and port over the other mods.  This saves a future PR and the associated review delays.

**Comment by pchote** [comment]
Updated.  This now updates all our default mods and removes support for the old syntax.

This will temporarily break downstream mods that depend on stacked upgrades, but that will be fixed by the time the migration is complete.  The alternative would require more PRs and reviewing overhead.

File: OpenRA.Game/FieldLoader.cs [hunk 3]
```diff
@@ -396,6 +398,22 @@ public static object GetValue(string fieldName, Type fieldType, MiniYaml yaml, M
 
 				return InvalidValueAction(value, fieldType, fieldName);
 			}
+			else if (fieldType == typeof(BooleanExpression))
+			{
+				if (value != null)
```
Comment (Phrohdoh) [inline OpenRA.Game/FieldLoader.cs]:
```csharp
if (value == null)
    return ...
```

**Review by phrohdoh — DISMISSED** [review]
Looks like a great improvement to me and I didn't see any issues with cloaking and experience. 👍

File: OpenRA.Game/FieldLoader.cs (hunk 3 above)
Comment (pchote) [inline OpenRA.Game/FieldLoader.cs]: This is just following the convention of the rest of the method.

**Comment by GraionDilach** [comment]
>This will temporarily break downstream mods that depend on stacked upgrades, but that will be fixed by the time the migration is complete.

Please assign atleast the issue to next release then.

**Comment by pchote** [comment]
Done.

File: mods/ts/rules/defaults.yaml [hunk 4]
```diff
@@ -18,69 +18,80 @@
...
-	SpeedMultiplier@EXPERIENCE:
-		UpgradeTypes: speed
-		Modifier: 120, 140
-	ReloadDelayMultiplier@EXPERIENCE:
-		UpgradeTypes: reload
-		Modifier: 90, 75
+			500: rank-veteran
+			1000: rank-elite
+	FirepowerMultiplier@VETERAN:
+		RequiresCondition: (rank-veteran && !rank-elite) || crate-firepower
```
Comment (GraionDilach) [inline mods/ts/rules/defaults.yaml]: While I'm aware that this is the sole usecase for combined expressions, could this be changed now to implement the TS crate multipliers - https://github.com/LipkeGu/TSIniFiles/blob/master/rules.ini#L10467 and onwards with the last coloumn being them?

**Comment by pchote** [comment]
I have found an issue with traits that rely on two or more conditions.  Fixing this will require reworking the internal upgrade plumbing (which I was *really* hoping to put off to one of the later prs).

**Comment by pchote** [comment]
The problem I found was that the `UpgradeManager` code was giving bogus levels for expressions that had multiple conditions, following its original rules of treating multiple entries in `UpgradeTypes` as an OR.

I've included the core `UpgradeManager` code changes that I was originally planning for a later PR, keeping the old naming and shim methods in place for the upgrade-granting code that still needs to be ported.

The new code set set up to allow for serialization and also to work with traits that want to grant conditions by default. #12396 provides another test case for these latter features in action.

**Comment by pchote** [comment]
This now also removes the upgrades lint check until the conversion is complete.

The old lint check was validating the old system where you granted as many upgrades as you needed, but never ones that you didn't need, and then each trait can only care about one thing at a time (e.g. `notmobile` which is granted by N traits and consumed by anything that needs to know when the actor can't move).

The new system inverts this, having traits grant one specific condition per action, and then traits can combine these as they want on the consumer end (e.g. `waiting-for-pickup` on a carryable actor).

Right now we're stuck in the middle of these two approaches, and can't meaningfully validate *either*.  This gives us false positives that break the build.

**Comment by reaperrr** [comment]
I couldn't spot any issues in the code directly, but noticed two in-game regressions (there might be more) compared to bleed:
- turrets of the transported mammoth tanks on TD shellmap don't show up
- the buildings on the TS shellmap are not invulnerable anymore, same seems to be the case on the RA shellmap

**Comment by pchote** [comment]
Argh, wonderful.  That will be related to [this TODO/workaround](https://github.com/OpenRA/OpenRA/pull/12381/commits/d518632bb224658e7b044751230d2c8f5ed329eb#diff-321a14c48088cce630500a3416764030R118).
I'm probably going to need to pull in even more code changes to solve that properly.

Maybe I should give up on trying to split this, and have one monster PR that does everything (hopefully) actually works...

**Comment by reaperrr** [comment]
> Maybe I should give up on trying to split this, and have one monster PR that does everything and (hopefully) actually works...

I think finding all bugs or regressions (if there are any) before anything is merged is extremely hard with changes of this magnitude either way.

**Comment by pchote** [comment]
Fixed.  That wasn't as bad as I was expecting.

* Fixed missing turrets by [correcting the fallback value for EnabledByDefault to true](https://github.com/OpenRA/OpenRA/pull/12381/commits/05187f5828310a5bac2d50fc8c89d4f5a4812180#diff-3103566713d363c66f2015400979d6edR37).
* Fixed the shellmaps by [initializing all known conditions](https://github.com/OpenRA/OpenRA/pull/12381/commits/3f0b33992c03a20ef0b26aa50b54a21e78f19053#diff-321a14c48088cce630500a3416764030R105) so that [AcceptsUpgrade](https://github.com/OpenRA/OpenRA/pull/12381/commits/3f0b33992c03a20ef0b26aa50b54a21e78f19053#diff-321a14c48088cce630500a3416764030R239) didn't throw out known-but-false conditions.
* Changed `Cargo` and `Carryable` to query `UpgradeManager` from `INotifyCreated.Created` so that `UpgradeManager` can `Requires<IConditionConsumerInfo>` to guarantee initialization order.

**Review by reaperrr — APPROVED** [review]
The only remaining issue I noticed was with a TS carryall carrying a TickTank, but that's fixed by #12396, so 👍

**Comment by pchote** [comment]
The Carryall/TickTank issue has always been there, and is one of the main motivations for this rework.

**Review by GraionDilach — APPROVED** [review]
Things work fine here - at one point TS eliteweapon antics did threw me off but that's Westwood and has nothing to do with this PR.

 :+1:

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9d51402c80f0e1a1254e35b49a5b299e5ea5600f...455d8a5639323686751636c1f7e6dd2de41d4739)

**Pull request merged**